controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -345,21 +360,106 @@ public void deleteStream(StreamInfo request, StreamObserver<DeleteStreamStatus>
+
+            StreamAuthParams authParams = new StreamAuthParams(request.getScope(), request.getStream(),
+                    request.getRequestedPermission(), this.isInternalWritesWithReadPermEnabled);
+
+            // StreamResource will be a stream representation (ex: "prn:://scope:myScope/stream:_RGmyApp") of the
+            // reader group (ex: "prn:://scope:myScope/reader-group:myApp). We use stream representation in claims
+            // put in delegation tokens for Segment Store's use, even though we use the regular representation for
+            // authorization here in the Controller.
+            String streamResource = authParams.streamResourceString();
+            String resource = authParams.resourceString();
+
+            if (authParams.isRequestedPermissionEmpty()) {
+                // For backward compatibility: Older clients will not populate requested permission.
+                log.info("Requested permission was empty for request with scope {} and stream {}", request.getScope(),
+                        request.getStream());
+                return this.grpcAuthHelper.checkAuthorizationAndCreateToken(resource, AuthHandler.Permissions.READ_UPDATE);
+            } else {
+                log.debug("Requested permission was {} for request with scope {} and stream {}",
+                        request.getRequestedPermission(), request.getScope(), request.getStream());
+
+                // The resource string that'll be used in the delegation token for use of the segment store
+                final String tokenResource;
+
+                // The operation that'll be specified as granted for the resource in the token. The bearer of the token
+                // will be allowed to perform the specified operation.
+                final AuthHandler.Permissions tokenPermission;
+
+                // This is the permission that the client is requesting to be assigned on the delegation token.
+                AuthHandler.Permissions requestedPermissions = authParams.requestedPermission();
+
+                if (authParams.isStreamUserDefined()) {
+                    // The operation itself requires the caller to possess read permissions.
+                    AuthHandler.Permissions minimumPermissions = AuthHandler.Permissions.READ;
+
+                    if (requestedPermissions.equals(AuthHandler.Permissions.READ_UPDATE) ||
+                            requestedPermissions.equals(minimumPermissions)) {
+                        this.grpcAuthHelper.checkAuthorization(streamResource, requestedPermissions);
+                        tokenResource = streamResource;
+                        tokenPermission = requestedPermissions;
+                    } else {
+                        // The minimum permission that the user must have to be able to invoke this call. This
+                        // authorizes the operation.
+                        this.grpcAuthHelper.checkAuthorization(streamResource, minimumPermissions);
```

Comment:
i am sorry, i think i had asked this earlier too - why do we authorize for both minimum permissions and requested permissions successively.. isnt minimum permission redundant in that case?

Reply:
The minimum permissions is the permission required to make this call. The requested permission is the permission the caller wants to be assigned in the delegation token. 

In most cases, the calling code from the client will ask for one of these: 
* READ - in which case authorization is only done for READ
* READ_WRITE or WRITE - in which case authorization is only done for READ_UPDATE
* No permission - in which case the default/backward compatibility rule applies. 

This particular piece of code is for other unexpected cases. Here's an example: https://github.com/pravega/pravega/blob/ed3f57f786084ae07c2657d4134507754859e0a5/controller/src/test/java/io/pravega/controller/server/rpc/grpc/v1/InMemoryControllerServiceImplTest.java#L170